airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryWriter.java

```diff
@@ -0,0 +1,268 @@
+    } catch (final ConfigNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  public void writeDestinationConnection(final DestinationConnection destination, final ConnectorSpecification connectorSpecification)
+      throws JsonValidationException, IOException {
+    final var previousDestinationConnection = getOptionalDestinationConnection(destination.getDestinationId())
+        .map(DestinationConnection::getConfiguration);
+
+    final JsonNode partialConfig = statefulUpdateSecrets(
+        destination.getWorkspaceId(),
+        previousDestinationConnection,
+        destination.getConfiguration(),
+        connectorSpecification);
+    final DestinationConnection partialDestination = Jsons.clone(destination).withConfiguration(partialConfig);
+
+    // validate partial to avoid secret leak issues.
+    validator.ensure(connectorSpecification.getConnectionSpecification(), partialDestination.getConfiguration());
+
+    configRepository.writeDestinationConnectionNoSecrets(partialDestination);
+  }
+
+  /**
+   * Detects secrets in the configuration. Writes them to the secrets store. It returns the config
+   * stripped of secrets (replaced with pointers to the secrets store).
+   *
+   * @param workspaceId workspace id for the config
+   * @param fullConfig full config
+   * @param spec connector specification
+   * @return partial config
+   */
+  public JsonNode statefulSplitSecrets(final UUID workspaceId, final JsonNode fullConfig, final ConnectorSpecification spec) {
+    return splitSecretConfig(workspaceId, fullConfig, spec, longLivedSecretPersistence);
+  }
+
+  /**
+   * @param workspaceId workspace id for the config
+   * @param oldConfig old full config
+   * @param fullConfig new full config
+   * @param spec connector specification
+   * @return partial config
+   */
+  public JsonNode statefulUpdateSecrets(final UUID workspaceId,
+                                        final Optional<JsonNode> oldConfig,
+                                        final JsonNode fullConfig,
+                                        final ConnectorSpecification spec) {
+    if (longLivedSecretPersistence.isPresent()) {
+      if (oldConfig.isPresent()) {
+        final var splitSecretConfig = SecretsHelpers.splitAndUpdateConfig(
+            workspaceId,
+            oldConfig.get(),
+            fullConfig,
+            spec,
+            longLivedSecretPersistence.get());
+
+        splitSecretConfig.getCoordinateToPayload().forEach(longLivedSecretPersistence.get()::write);
+
+        return splitSecretConfig.getPartialConfig();
+      } else {
+        final var splitSecretConfig = SecretsHelpers.splitConfig(
+            workspaceId,
+            fullConfig,
+            spec);
+
+        splitSecretConfig.getCoordinateToPayload().forEach(longLivedSecretPersistence.get()::write);
+
+        return splitSecretConfig.getPartialConfig();
+      }
+    } else {
+      return fullConfig;
```

Comment:
Does this have the potential to break the guarantee stated in the javadoc?
> never returns a secrets as return values (even the ones that are passed in as arguments)

Reply:
fair. moved this method to be `private`. added a todo to get rid of the case where a secret can be returned. the root problem here is that we handle secrets so differently when a user does or does not provide a secrets store. fixing that is out of scope for this PR, but once we fix that, then we can get get the guarantee we are looking for.